frame/staking/src/lib.rs

```diff
@@ -654,6 +753,23 @@ pub trait Trait: frame_system::Trait {
 	/// The NPoS reward curve to use.
 	type RewardCurve: Get<&'static PiecewiseLinear<'static>>;
 
+	/// Something that can estimate the next session change, accurately or as a best effort guess.
+	type NextSessionChange: EstimateNextSessionChange<Self::BlockNumber>;
```

Comment:
I think this trait is confusing, what next session change are we talking about ? is it:
* in how many block does `SessionManager::new_session(current_era + 1)` will be called ?
* in how many block does `SessionManager::start_session(current_era + 1)` will be called ?

because we actually want the former but Babe implements the latter

Reply:
This is a very important comment. Session has changed a bit and I need to make sure it is correct.

So we want the session change at which the election should happen, so we can trigger the offchain flag n block before it. 

Can you explain what is the difference between the two?

Reply:
here documentation of SessionManager:
```rust
/// A trait for managing creation of new validator set.
pub trait SessionManager<ValidatorId> {
	/// Plan a new session, and optionally provide the new validator set.
	///
	/// Even if the validator-set is the same as before, if any underlying economic
	/// conditions have changed (i.e. stake-weights), the new validator set must be returned.
	/// This is necessary for consensus engines making use of the session module to
	/// issue a validator-set change so misbehavior can be provably associated with the new
	/// economic conditions as opposed to the old.
	/// The returned validator set, if any, will not be applied until `new_index`.
	/// `new_index` is strictly greater than from previous call.
	///
	/// The first session start at index 0.
	fn new_session(new_index: SessionIndex) -> Option<Vec<ValidatorId>>;
	/// End the session.
	///
	/// Because the session pallet can queue validator set the ending session can be lower than the
	/// last new session index.
	fn end_session(end_index: SessionIndex);
	/// Start the session.
	///
	/// The session start to be used for validation
	fn start_session(start_index: SessionIndex);
}
```

So new_session is called to retrieve a potential new validator set for a future session, if a session implements no delay, it would call `new_session(0), start_session(0), end_session(0), new_session(1), start_session(1), end_session(1)...`
but a module could also do:
`new_session(0), new_session(1), start_session(0), end_session(0), new_session(2), start_session(1), end_session(1)...`
which is what session-pallet does.

but it could also do:
```
new_session(0);new_session(1);new_session(2);new_session(3);
start_session(0); end_session(0);
start_session(1); end_session(1);
start_session(2); end_session(2);
start_session(3); end_session(3);
new_session(4); new_session.....
```

I mean the trait itself does not give any constraint.

So what we want to have is to know an estimation of the block number of the next call to next_session. Babe does not have this information as babe is responsible to trigger the start of a new session and has no knowledge about when its validator set has actually been choose.

I'm not sure how to fix it, maybe we can put more constrain on SessionManager but I don't see how, or we create a trait EstimateNextNewSession which is implemented by pallet-session.

Reply:
Thanks for the explanation! 

Maybe we can move `EstimateNextNewSession` to Session, but it would be the same logic proxied behind session, since session itself also doesn't have any knowledge of the length of the session and uses 
https://github.com/paritytech/substrate/blob/75116bd8c45c7e46a9b8eed9cb29ee4256ed631b/frame/session/src/lib.rs#L510-L514
which is essentially `Babe`. It would actually only be less efficient since then we have to roughly do: 

```
impl<T> EstimateNextNewSession for session::Module<T> {
  fn estimate(now: u32) -> u32 {
     n = now 
     while ! T::ShouldEndSession::should_end_session(n) { n ++ }
    n
  }
}
```

> So what we want to have is to know an estimation of the block number of the next call to next_session. **Babe does not have this information as babe is responsible to trigger the start of a new session and has no knowledge about when its validator set has actually been choose**.

I don't see this in the code though. Babe triggers the entire `Self::rotate_session` and within `Self::rotate_session` we call into both `new_session` and `start_session`. So in some sense Babe is the overarching entity that dictates the session at the end of the day.

Reply:
Yes I think having a trait EstimateNextNewSession implemented by pallet-session which calls inside EstimateNextShouldEndSession implemented by babe seems to work

> I don't see this in the code though. Babe triggers the entire Self::rotate_session and within Self::rotate_session we call into both new_session and start_session. So in some sense Babe is the overarching entity that dictates the session at the end of the day.

This is a specific implement of SessionManager by the current pallet-session nothing specified in the trait constraints.

Reply:
But what I proposed is unnecessarily more expensive: there is a loop which checks all block from now up until when `T::ShouldEndSession` returns true, while if we let babe do it babe can do it without this loop.

Reply:
We consolidated this offline: 

`Babe` and `PerdiodicSessionChange` implement `EstimateNextSessionRotatio`. Session has access to this via. `type NextSessionRotation: EstimateNextSessionRotatio`. from the commets: 

```rust
/// Something that can predict the next session rotation. This should typically come from the
/// same logical unit that provides [`ShouldEndSession`], yet, it gives a best effort estimate.
/// It is helpful to implement [`EstimateNextNewSession`].
type NextSessionRotation: EstimateNextSessionRotation<Self::BlockNumber>;
```

And then Session is the only thing that implements `EstimateNextNewSession`, which just calls into `T::NextSessionRotation`. But as mentioned by @thiolliere the point of this is that the current session module schedules a new session and starts the previous one at one go all at the same time, in `fn rotate_session`. A different session module can customise this if needed. 

Staking finally accepts `type NextNewSession: EstimateNextNewSession` which is implemented by Session. So Staking is no loner dependent on Babe but rather Session. 

Pushed in a few minutes.